borg/repository.py

```diff
@@ -591,6 +602,28 @@ def get_fd(self, segment):
             self.fds[segment] = fd
             return fd
 
+    def _sync_and_close(self, fd):
+        fd.flush()
+        os.fdatasync(fd.fileno())
+        fd.close()
+
+    def close_segment(self, sync=False):
+        if len(self._write_fds) == self.MAX_PENDING_WRITE_FDS:
+            self._sync_and_close(self._write_fds.popleft())
+        if self._write_fd:
+            self.segment += 1
+            self.offset = 0
+            self._write_fds.append(self._write_fd)
+            if hasattr(os, 'posix_fadvise'):  # only on UNIX
+                # tell the OS that it does not need to cache what we just wrote,
+                # avoids spoiling the cache for the OS and other processes.
+                os.posix_fadvise(self._write_fd.fileno(), 0, 0, os.POSIX_FADV_DONTNEED)
```

Comment:
So, make max_pending_write_fds much smaller (say, 30 or so)?

Anyway, it's interesting that fadvise doesn't work / is said to not work. From the other stuff that I read lately and what verygreen said I can only conclude that *adivse are utterly broken and misleading APIs.